lib/entities/HardQuoteRequest.ts

```diff
@@ -0,0 +1,113 @@
+import { TradeType } from '@uniswap/sdk-core';
+import { UnsignedV2DutchOrder } from '@uniswap/uniswapx-sdk';
+import { BigNumber, utils } from 'ethers';
+
+import { HardQuoteRequestBody } from '../handlers/hard-quote';
+import { QuoteRequestDataJSON } from '.';
+
+export class HardQuoteRequest {
+  public order: UnsignedV2DutchOrder;
+
+  public static fromHardRequestBody(_body: HardQuoteRequestBody): HardQuoteRequest {
+    // TODO: parse hard request into the same V2 request object format
+    throw new Error('Method not implemented.');
+  }
+
+  constructor(private data: HardQuoteRequestBody) {
+    this.order = UnsignedV2DutchOrder.parse(data.encodedInnerOrder, data.tokenInChainId);
+  }
+
+  public toCleanJSON(): QuoteRequestDataJSON {
+    return {
+      tokenInChainId: this.tokenInChainId,
+      tokenOutChainId: this.tokenOutChainId,
+      swapper: utils.getAddress(this.swapper),
+      requestId: this.requestId,
+      tokenIn: this.tokenIn,
+      tokenOut: this.tokenOut,
+      amount: this.amount.toString(),
+      type: TradeType[this.type],
+      numOutputs: this.numOutputs,
+      ...(this.quoteId && { quoteId: this.quoteId }),
+    };
+  }
+
+  // return an opposing quote request,
+  // i.e. quoting the other side of the trade
+  public toOpposingCleanJSON(): QuoteRequestDataJSON {
+    const type = this.type === TradeType.EXACT_INPUT ? TradeType.EXACT_OUTPUT : TradeType.EXACT_INPUT;
+    return {
+      tokenInChainId: this.tokenInChainId,
+      tokenOutChainId: this.tokenOutChainId,
+      requestId: this.requestId,
+      swapper: utils.getAddress(this.swapper),
+      // switch tokenIn/tokenOut
+      tokenIn: utils.getAddress(this.tokenOut),
+      tokenOut: utils.getAddress(this.tokenIn),
+      amount: this.amount.toString(),
+      // switch tradeType
+      type: TradeType[type],
+      numOutputs: this.numOutputs,
+      ...(this.quoteId && { quoteId: this.quoteId }),
```

Comment:
should we change quoteId? Otherwise first quoteId = real quote?

Reply:
keeping same semantics as the v1 quote, interesting point tho - I think quoters can disintermediate it themselves by storing both quotes under same quoteId and assuming only one or the other will end up as an order